### DISCUSSION TITLE
Return client's own defined roles in access token for grant type of client_credentials

### DISCUSSION THREAD

**Discussion by idavollen** [keycloak/keycloak#30429] [Q&A] [answered]
We'are currently using the grant type of _password_ to secure our external APIs where the resource owner and the client are the same party. If I define some roles for a particular client_id, those client's tailored roles will be returned for a certain username associated with that particular client_id by configuring **Role Mappings** for that specific user on the Keycloak Admin Console. 

However, the grant type of _password_ will be deprecated in OAuth2.1, which means that we'd have to migrate to the grant type of _client_credentials_ to secure our external API. It seems that there is no way to return those client tailored roles in access token when the grant type of _client_credentials_ is being used. 

I'm wondering if Keycloak will provide backwards compatibility of grant type of _password_ when OAuth2.1 is to be widely used? or if the client's tailored roles could be returned in the access token for the grant type of _client_credentials_ ?

**Comment by bonnm** [comment]
In the client settings, go to "Client scopes" tab, select the "...-dedicated" scope, then add the predefined mappers "client roles", "audience resolve" and "realm roles" (maybe the last two can be skipped). Then, in the Scope tab, assign the client roles you want to add.

That should do it.

**Reply by idavollen** [reply]
@bonnm thanks for your reploy! We're using outdated _redhat-sso73-openshift_ (Keycloak) in my company, I can't manage to find those mentioned GUI parts

**Comment by bonnm** [comment]
I'm not sure, but I think that was also possible with old Keycloak Versions:
https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/7.3/html/server_administration_guide/clients#doc-wrapper

**Comment by idavollen** [comment]
![image](https://github.com/keycloak/keycloak/assets/1467270/3103a808-d98c-43e3-b95a-f84e9ce84fa9)
that is how the Client Scopes look like with us

**Comment by bonnm** [comment]
Take the "Mappers" tab and the "Scope" tab...

**Comment by idavollen** [comment]
![image](https://github.com/keycloak/keycloak/assets/1467270/e812f710-92d7-4112-9ab1-7fd7fbd15f26)
I've tried to **Add Builtin** _client roles_, however, those client roles will be returned in access token only for the grant type of _password_, but not for the grant type of _client_credentials_

**Comment by idavollen** [comment]
The Scope tab has only one setting: **Full Scope Allowed**: **ON**

**Comment by idavollen** [comment]
![image](https://github.com/keycloak/keycloak/assets/1467270/4797cb8a-273b-4935-808e-08838204879f)
After enabling this setting, the builtin role _uma_protection_ is automatically added under the Client -> Roles tab, where I tried to add my tailored roles, but _only_ the builtin _uma_protection_ was returned, which is really strange.
![image](https://github.com/keycloak/keycloak/assets/1467270/b37b32d2-db45-4fc0-bce0-6b427e3b1b95)
![image](https://github.com/keycloak/keycloak/assets/1467270/6210f910-a4e1-4a89-b98d-4f618c0089b6)

**Comment by bonnm** [comment]
Try setting the full scope allowed to off, then the GUI will change to something where you can add/remove needed rolens from a list. Add the needed client roles. If that is not enought, try adding also the "realm roles" and "audience resolve" mappers.

The "Authorization Enabled" switch is definiely not necessary,

https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/7.3/html/server_administration_guide/clients#service_accounts

**Reply by idavollen** [reply]
I have turned off _Full Scope Allowed_, and also disabled the "**Authorization Enabled**", 
![image](https://github.com/keycloak/keycloak/assets/1467270/26efb9dc-f3ed-497c-a401-335512fc4a92)
and the access token contains only the builtin _uma_protection_ for the grant type of _client_credentials_, but those client's own defined roles are not included.
![image](https://github.com/keycloak/keycloak/assets/1467270/f3c04b45-97e7-4e47-9e85-9fed51c1d674)

**Comment by idavollen** [comment]
What I hope to achieve is to get those marked **_Effective Roles_** on this above mentioned image returned in the access token

**Comment by bonnm** [comment, marked as answer]
In the tab "Service Account Roles", have you added the desired roles to the service account? Without that, des client roles will just work for a normal user with password.

**Reply by idavollen** [reply]
@bonnm it works as what I want. Really appreciate your professional help!

**Reply by TeresaMarie43** [reply]
Some help would be great thanks
